Allow file:// clones of submodules in tests

MarkEWaite opened the pull request:
## Allow file:// clones of submodules in tests

Allow local git clones to use the file:// protocol by setting protocol.file.allow=always in the global git configuration for the current user.

Command line git 2.38.1 and patches to earlier versions disallow local git submodule cloning with the file:// protocol. The change resolves a security issue but that security issue is not a threat to these tests.

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/jenkinsci/git-client-plugin/blob/master/CONTRIBUTING.adoc) doc
- [x] I have added tests that verify my changes
- [x] Unit tests pass locally with my changes
- [x] I have added documentation as necessary
- [x] No Javadoc warnings were introduced with my changes
- [x] No spotbugs warnings were introduced with my changes

## Types of changes

- [x] Test improvement

jglick commented on src/test/java/org/jenkinsci/plugins/gitclient/GitClientTest.java:
```diff
@@ -258,6 +259,32 @@ public void setGitClient() throws IOException, InterruptedException {
+     * Removes the protocol.file.allow setting from the global git
+     * configuration of the current user.
+     */
+    @After
```
Better to factor this out into a common method, even a `TestRule`.

MarkEWaite replied:
I'll see if I can find a way to factor it into a common location.  

I've never written a `TestRule` before and I'm running out of weekend time that I can invest, so if a `TestRule` is needed, then I'll likely need to wait a few weeks before I can address the issue.  Next weekend and the following weekend will be filled with watching the grandchildren while their parents help a Lego League event.

jglick replied:
https://junit.org/junit4/javadoc/4.12/org/junit/rules/ExternalResource.html is the easiest example. Just a suggestion of an idiomatic way to remove the duplicated code.

jglick commented on src/test/java/org/jenkinsci/plugins/gitclient/GitAPITestUpdate.java:
```diff
@@ -439,6 +438,58 @@ public void tearDown() throws Exception {
+    }
+
+    /**
+     * Removes the protocol.file.allow setting from the global git
```
What if the developer’s Git config deliberately included this flag before running tests?

MarkEWaite replied:
This will remove the flag even if it was previously set.  I didn't feel that was a critical condition to preserve in the developer's git config.  If you feel that it is, I'm happy to investigate some way of detecting the current state and preserving the current state.  I've not been very successful with those techniques in the past, but I'm willing to try again if it is a serious concern.

MarkEWaite replied:
Resolved by switching from a change of global configuration to changing the arguments to command line git.  Thanks for the suggestion!

jglick reviewed (commented):
This does not seem like a good idea. What if the test JVM crashes? Etc.

Acc. to https://git-scm.com/docs/git-config#FILES you can set `XDG_CONFIG_HOME` at least on Unix.

You could also just fix `git-client` to pass `-c protocol.file.allow=always` when some internal flag is set, which we would set during tests.

jglick commented:
(BTW when you force-push to a PR branch you break GH’s review flow, making it impossible to see at a glance what is new vs. what I already reviewed.)

jglick commented:
In the longer term, I would suggest redesigning CLI Git tests to use testcontainers.org so that tests run in a well-defined environment, rather than suddenly breaking the moment someone happens to update the Git version used in CI machines for purposes of plugin source code checkout. I was in the process of a similar conversion in `mercurial-plugin` when I worked on it.

MarkEWaite commented:
> (BTW when you force-push to a PR branch you break GH’s review flow, making it impossible to see at a glance what is new vs. what I already reviewed.)

Sorry about that.  I had hoped that I had "won the race" and force pushed before you started reviewing.  Will make sure that I don't force push in the future.  Thanks for the review comments.

I like the idea of adding an extra argument to command line git calls in these special cases.  Would completely avoid changes to global git configuration.

MarkEWaite commented:
@jglick I've simplified the code significantly and switched to using a command line modifier rather than a global configuration modifier.  Once the tests are successful, it is ready for a review.

jglick commented on src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java:
```diff
@@ -2314,6 +2314,41 @@ void fixSshKeyOnWindows(Path key) throws GitException {
+     * The change resolves a security issue but that security issue is
+     * not a threat to these tests.
+     *
+     * NOT INTENDED FOR USE OUTSIDE THE PLUGIN.
```
Clear enough from lack of `public` modifier, but OK.

jglick commented on src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java:
```diff
@@ -2648,6 +2683,10 @@ private String launchCommandIn(ArgumentListBuilder args, File workDir, EnvVars e
         }
+        /* Prepend extra git command line arguments if any */
+        if (!extraGitCommandArguments.isEmpty()) {
+            args = args.prepend(extraGitCommandArguments.stream().toArray(String[]::new));
```
or more simply (and [faster](https://stackoverflow.com/a/55414113/12916)!)
```suggestion
            args = args.prepend(extraGitCommandArguments.toArray(new String[0]));
```

jglick commented on src/test/java/org/jenkinsci/plugins/gitclient/GitAPITestUpdate.java:
```diff
@@ -258,12 +257,21 @@ String contentOf(String path) throws IOException {
             return FileUtils.readFileToString(file(path), "UTF-8");
         }
 
+        private CliGitAPIImpl cachedCliGitAPIImpl = null;
```
Why are you bothering with a cache of a flyweight object in a unit test?

MarkEWaite replied:
> Why are you bothering with a cache of a flyweight object in a unit test?

There are tests that fail unless the same cgit object is returned for multiple calls to a WorkingArea object.  The case that inspired it was JGit as the tested implementation, but the test needs to use CGit features to create submodules before exercising JGit to confirm the submodule checkout works as expected.  The caching makes the change "sticky".

I can add that information as a comment on the cache so that my future self and others can read the rationale in the source code.

jglick replied:
Yes a comment would be helpful, since the term “cache” is usually understood as solely a performance optimization.

MarkEWaite replied:
Done in 0a828e3f3caf241efe3b16d51d5038a258035c36

jglick commented on src/test/java/org/jenkinsci/plugins/gitclient/GitAPITestUpdate.java:
```diff
@@ -924,10 +931,10 @@ public void testGetSubmodules() throws Exception {
     @Test
     public void testSubmoduleUpdateShallow() throws Exception {
         WorkingArea remote = setupRepositoryWithSubmodule();
+        w.cgit().allowFileProtocol(); // CLI git 2.38.1 requires protocol.file.allow=always for this test
```
I think the comment can be dropped here—Javadoc of the method being called explains this in detail.

MarkEWaite replied:
Done in 7f8b9879477b0fd68b667720e9cbf46237e20dbe

jglick commented on src/test/java/org/jenkinsci/plugins/gitclient/GitClientTest.java:
```diff
@@ -258,6 +256,23 @@ public void setGitClient() throws IOException, InterruptedException {
+     * Allow local git clones to use the file:// protocol by setting
+     * protocol.file.allow=always on the git command line of the
+     * GitClient argument that is passed.
+     *
```
```suggestion
     * <p>
```

for anyone actually reading as HTML

jglick reviewed (approved):
Glad to see that this worked!
